src/hooks/balances/lib/getBalancesByAddress.tsx

```diff
@@ -0,0 +1,87 @@
+import { ApiPromise } from '@polkadot/api';
+import { includes } from 'lodash';
+import { Balance } from '../../../generated/graphql';
+import constants from '../../../constants';
+import { OrmlAccountData } from '@open-web3/orml-types/interfaces';
+import '@polkadot/api-augment';
+import { FrameSystemAccountInfo } from '@polkadot/types/lookup';
+
+// this function fetches asset balances only for a given set of assetIds
+export const getBalancesByAddress = async (
+  apiInstance: ApiPromise,
+  address: string,
+  assetIds: string[]
+): Promise<Balance[]> => {
+  let balances: Balance[] = [];
+
+  if (includes(assetIds, constants.nativeAssetId)) {
+    const nativeBalance = await fetchNativeAssetBalance(apiInstance, address);
+    balances.push(nativeBalance);
+  }
+
+  // make sure that there is no native assetId
+  const nonNativeAssetIds = assetIds.filter(
+    (id) => id !== constants.nativeAssetId
+  );
+  // fetch non-native assets only if needed
+  if (nonNativeAssetIds.length) {
+    const nonNativeBalances = await fetchNonNativeAssetBalances(
+      apiInstance,
+      address,
+      nonNativeAssetIds
+    );
+
+    balances.push(...nonNativeBalances);
+  }
+
+  return balances;
+};
+
+export const fetchNativeAssetBalance = async (
+  apiInstance: ApiPromise,
+  address: string
+): Promise<Balance> => {
+  // no handling of undefined because apiInstance returns default value of 0 for native asset
+  const nativeAssetBalance: FrameSystemAccountInfo =
+    await apiInstance.query.system.account(address);
+
+  return {
+    assetId: constants.nativeAssetId,
+    balance: nativeAssetBalance.data.free.toString(),
+  };
+};
+
+/**
+ * This function fetches balances for multiple non-native assetIds.
+ *
+ * @param apiInstance PolkadotJS ApiPromise
+ * @param address of the account
+ * @param assetIds of non-native tokens
+ * @returns balance object with assetId and balance
+ */
+export const fetchNonNativeAssetBalances = async (
+  apiInstance: ApiPromise,
+  address: string,
+  assetIds: string[]
+): Promise<Balance[]> => {
+  // compose search parameter as [[address, assetIdA], [address, assetIdB], ...]
+  const queryParameter: string[][] = assetIds.map((assetId) => [
+    address,
+    assetId,
+  ]);
+  const searchResult =
+    await apiInstance.query.tokens.accounts.multi<OrmlAccountData>(
+      queryParameter
+    );
+
+  const balances: Balance[] = searchResult.map((balanceData, i) => {
+    // extract free balance as string
+    const freeBalance = balanceData.free.toString();
+
+    return {
+      assetId: assetIds[i], // pair assetId in the same order as provided in query parameter
```

Comment:
I am not entirely sure about this, you're relying on the fact that `.multi` returns data in the same order as you've requested it.

This could have far reaching consequences as we're using this to determine e.g. the spot price for a pool.

I haven't found anything in the docs that would mend this, e.g. pairing of the parameters provided with the results returned. Let's keep it as it is for now.

Reply:
I agree that we have to watch out for this. I also did not find anything but we need to tell QA/testing.